hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/preloader/NearCachePreloader.java

```diff
@@ -0,0 +1,244 @@
+     *
+     * @param records the {@link NearCacheRecordMap} instances to retrieve the keys from
+     * @param <R>     the {@link NearCacheRecord} type
+     * @param <NCRM>  the {@link NearCacheRecordMap} type
+     */
+    public <R extends NearCacheRecord, NCRM extends NearCacheRecordMap<KS, R>> void storeKeys(NCRM... records) {
+        long started = System.nanoTime();
+        FileOutputStream fos = null;
+        try {
+            lastWrittenBytes = 0;
+            lastKeyCount = 0;
+
+            fos = new FileOutputStream(tmpStorageFile, false);
+            for (NCRM record : records) {
+                writeKeySet(fos, fos.getChannel(), record.keySet());
+            }
+            if (lastKeyCount == 0) {
+                deleteQuietly(storageFile);
+                return;
+            }
+            fos.flush();
+
+            rename(tmpStorageFile, storageFile);
+        } catch (Exception e) {
+            logger.warning(format("Could not store keys of Near Cache %s: %s", nearCacheName, e.getMessage()));
+        } finally {
+            deleteQuietly(tmpStorageFile);
+            closeResource(fos);
+
+            long elapsedMillis = getElapsedMillis(started);
+            nearCacheStats.addPersistence(elapsedMillis, lastWrittenBytes, lastKeyCount);
+
+            logger.info(format("Stored %d keys of Near Cache %s in %d ms (%d kB)", lastKeyCount, nearCacheName, elapsedMillis,
+                    lastWrittenBytes / KB_FACTOR));
+        }
+    }
+
+    private int loadKeySet(BufferingInputStream bis, DataStructureAdapter<Data, ?> adapter) throws IOException {
+        byte[] sizeBytes = new byte[INT_SIZE_IN_BYTES];
+        int loadedKeys = 0;
+
+        Builder<Data> builder = InflatableSet.newBuilder(LOAD_BATCH_SIZE);
+        while (readFullyOrNothing(bis, sizeBytes)) {
+            int dataSize = toInt(sizeBytes);
+            byte[] payload = new byte[dataSize];
+            if (!readFullyOrNothing(bis, payload)) {
+                break;
+            }
+            builder.add(new HeapData(payload));
+            if (builder.size() == LOAD_BATCH_SIZE) {
+                adapter.getAll(builder.build());
+                builder = InflatableSet.newBuilder(LOAD_BATCH_SIZE);
+            }
+            loadedKeys++;
+        }
+        if (builder.size() > 0) {
+            adapter.getAll(builder.build());
+        }
+        return loadedKeys;
+    }
+
+    private void writeKeySet(FileOutputStream fos, FileChannel outChannel, Set<KS> keySet) throws IOException {
+        for (KS key : keySet) {
+            Data dataKey = serializationService.toData(key);
+            if (dataKey != null) {
+                int dataSize = dataKey.totalSize();
+
+                ensureBufHasRoom(fos, INT_SIZE_IN_BYTES);
+                buf.putInt(dataSize);
+
+                int position = 0;
+                int remaining = dataSize;
+                while (remaining > 0) {
```

Comment:
In cases where `remaining > BUFFER_SIZE` we can skip writing to the `ByteBuffer`, just as in `ChunkFileOut`.